### PR TITLE
Add initial support for FP8 ONNX export

### PR DESCRIPTION
**Pull request by tcherckez-nvidia**
Add support for ONNX_NAMESPACE::TensorProto_DataType_FLOAT8E5M2 and ONNX_NAMESPACE::TensorProto_DataType_FLOAT8E4M3FN to enable export of torch models that use FP8 (E4M3 and E5M2) to ONNX (opset 19)

### CODE REVIEW

**Review by thiagocrepaldi — COMMENTED** [review]
Thank you for your first contribution!

Could you add unit tests for each fp8 scenario? Also add a nice description for the PR.
Also, go to the `EasyCLA` check below, click on `Details` and sign it to be allowed to merge a PR towards Pytorch repo

File: third_party/pybind11
Comment (BowenBao) [inline third_party/pybind11:1]: please revert submodule update if it is accidental.

**Comment by tcherckez-nvidia** [comment]
Thanks for the welcoming comment @thiagocrepaldi !
Added the (simple)tests, and updated the description to be a bit more detailed.

Let me know if something else is needed.

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (thiagocrepaldi) [inline test/onnx/test_pytorch_onnx_no_runtime.py:1206]: Let's try to do an assert that shows the fp8 type is properly exported and present in the graph and not removed or changed during export process.

```suggestion
            onnx.checker.check_model(model_proto_file.name)
            onnx_type = { torch.float8_e4m3fn: 17, torch.float8_e5m2: 19}  # From https://github.com/onnx/onnx/blob/main/onnx/onnx.proto3#L512-L521
            loaded_model = onnx.load_from_string(model_proto_file.name
            self.assertEqual(
            loaded_model.graph.input[0].type.tensor_type.data_type, onnx_type[fp8_dtype])
        )
```

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (thiagocrepaldi) [inline test/onnx/test_pytorch_onnx_no_runtime.py:1204]:
```suggestion
                Model(), x, model_proto_file.name, opset_version=19
```
Should it be opset 19?

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (thiagocrepaldi) [inline test/onnx/test_pytorch_onnx_no_runtime.py:1202]: Sometimes using `io.BytesIO()` make the test simpler as you don't have to worry about write permission on the filesystem, etc

```python
        f = io.BytesIO()
        torch.onnx.export(Model(), x, f)
        loaded_model = onnx.load_from_string(f.getvalue())
```

File: torch/_C/_onnx.pyi
Comment (thiagocrepaldi) [inline torch/_C/_onnx.pyi:27]: Could you take this chance and also add the other [2 missing fp8 types](https://github.com/onnx/onnx/blob/main/onnx/onnx.proto3#L512-L521)?

Should be little extra work and a full fp8 support would be available :)

```
    # From ONNX spec
    FLOAT8E4M3FN = 17;    // float 8, mostly used for coefficients, supports nan, not inf 
    FLOAT8E4M3FNUZ = 18;  // float 8, mostly used for coefficients, supports nan, not inf, no negative zero 
    FLOAT8E5M2 = 19;      // follows IEEE 754, supports nan, inf, mostly used for gradients
    FLOAT8E5M2FNUZ = 20;  // follows IEEE 754, supports nan, inf, mostly used for gradients, no negative zero
```

Comment (tcherckez-nvidia) [reply torch/_C/_onnx.pyi:27]: My plan was to add all flavors, but this [MR](https://github.com/pytorch/pytorch/commit/b64bd4a5dd9ee21d269997c4a28f6389870cd540) added only fp8e4m3fn and fp8e5m2 to torch, so currently we can support only this datatypes export

File: test/onnx/test_pytorch_onnx_no_runtime.py
Comment (tcherckez-nvidia) [reply test/onnx/test_pytorch_onnx_no_runtime.py:1202]: sure, I was doing that because I wanted to inspect the resulting onnx file with Netron, but there is no need for that now

File: torch/csrc/jit/serialization/export.cpp
Comment (thiagocrepaldi) [inline torch/csrc/jit/serialization/export.cpp:88]:
```suggestion
        9, // opset 19
```
Not entirely sure, but I think opset 19 uses ir 9? Check this out https://github.com/onnx/onnx/blob/main/onnx/onnx.proto3#L106-L108

**Comment by thiagocrepaldi** [comment]
image

**Comment by tcherckez-nvidia** [comment]
> image

I've signed it, waiting for someone from my org to approve it

**Comment by thiagocrepaldi** [comment]
Several tests failed on CI:

```bash
onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset_version_18_is_script_True_keep_initializers_as_inputs_True::test_batchnorm2d_norunningstats FAILED [0.0201s] [ 91%]
2023-08-03T19:51:28.8765757Z 
2023-08-03T19:51:28.8765863Z ==================================== RERUNS ====================================
2023-08-03T19:51:28.8766067Z _ TestONNXRuntime_opset_version_18_is_script_True_keep_initializers_as_inputs_True.test_batchnorm2d_norunningstats _
2023-08-03T19:51:28.8766158Z Traceback (most recent call last):
2023-08-03T19:51:28.8766414Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/utils.py", line 1665, in _export
2023-08-03T19:51:28.8766493Z     _C._check_onnx_proto(proto)
2023-08-03T19:51:28.8766634Z RuntimeError: Unrecognized attribute: axes for operator ReduceMean
2023-08-03T19:51:28.8766678Z 
2023-08-03T19:51:28.8766824Z ==> Context: Bad node spec for node. Name: /ReduceMean OpType: ReduceMean
2023-08-03T19:51:28.8766830Z 
2023-08-03T19:51:28.8766972Z The above exception was the direct cause of the following exception:
2023-08-03T19:51:28.8766978Z 
2023-08-03T19:51:28.8767073Z Traceback (most recent call last):
2023-08-03T19:51:28.8767461Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 1853, in print_repro_on_failure
2023-08-03T19:51:28.8767525Z     yield
2023-08-03T19:51:28.8767813Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 2363, in wrapper
2023-08-03T19:51:28.8767889Z     method(*args, **kwargs)
2023-08-03T19:51:28.8768046Z   File "onnx/test_pytorch_onnx_onnxruntime.py", line 3998, in test_batchnorm2d_norunningstats
2023-08-03T19:51:28.8768124Z     self.run_test(model, x)
2023-08-03T19:51:28.8768279Z   File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 165, in run_test
2023-08-03T19:51:28.8768345Z     _run_test(
2023-08-03T19:51:28.8768492Z   File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 132, in _run_test
2023-08-03T19:51:28.8768571Z     return run_model_test(
2023-08-03T19:51:28.8768728Z   File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 82, in run_model_test
2023-08-03T19:51:28.8768867Z     return verification.verify(*args, options=options, **kwargs)
2023-08-03T19:51:28.8769030Z   File "<@beartype(torch.onnx.verification.verify) at 0x7f22bfaa1d30>", line 351, in verify
2023-08-03T19:51:28.8769293Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/verification.py", line 875, in verify
2023-08-03T19:51:28.8769361Z     utils._export(
2023-08-03T19:51:28.8769611Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/utils.py", line 1667, in _export
2023-08-03T19:51:28.8769710Z     raise errors.CheckerError(e) from e
2023-08-03T19:51:28.8769893Z torch.onnx.errors.CheckerError: Unrecognized attribute: axes for operator ReduceMean
2023-08-03T19:51:28.8769899Z 
2023-08-03T19:51:28.8770038Z ==> Context: Bad node spec for node. Name: /ReduceMean OpType: ReduceMean
2023-08-03T19:51:28.8770044Z 
2023-08-03T19:51:28.8770181Z During handling of the above exception, another exception occurred:
2023-08-03T19:51:28.8770190Z 
2023-08-03T19:51:28.8770273Z Traceback (most recent call last):
2023-08-03T19:51:28.8770559Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 2363, in wrapper
2023-08-03T19:51:28.8770634Z     method(*args, **kwargs)
2023-08-03T19:51:28.8770778Z   File "/opt/conda/envs/py_3.8/lib/python3.8/contextlib.py", line 131, in __exit__
2023-08-03T19:51:28.8770883Z     self.gen.throw(type, value, traceback)
2023-08-03T19:51:28.8771195Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 1860, in print_repro_on_failure
2023-08-03T19:51:28.8771343Z     e.args = (e.args[0] + f"\n{repro_str}",) + e.args[1:]
2023-08-03T19:51:28.8771547Z TypeError: unsupported operand type(s) for +: 'RuntimeError' and 'str'
2023-08-03T19:51:28.8771768Z ----------------------------- Captured stdout call -----------------------------
```

You can test locally by running `python -m pytest -sv onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset_version_18_is_script_True_keep_initializers_as_inputs_True::test_batchnorm2d_norunningstats`, for example

**Comment by thiagocrepaldi** [comment]
Lint CI also failed. Run

```bash
make setup_lint
lintrunner init
lintrunner -a
```
to fix the code style issues

File: test/onnx/test_pytorch_onnx_onnxruntime.py
Comment (tcherckez-nvidia) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:51]: opset 18 wasn't enabled for these tests.

tests were failing because opset version was bumped to 19, which made these tests to run with opset 18.
I updated this to be `-2` to keep those test to run only with opset<18

Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:51]: There are the `skipIfUnsupportedMinOpsetVersion(min_opset_version)` and `skipIfUnsupportedMaxOpsetVersion(max_opset_version)` if the failing cases are just a few

Comment (tcherckez-nvidia) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:51]: I'm familiar with those decorators, but I wish to keep the previous setup instead of making big changes.
enabling opset18 to run on my dev machine got ~250 failures, many of these related to torch not supporting the export of a specific op at all, or not complying to opset18 spec

**Review by thiagocrepaldi — COMMENTED** [review]
Running CI one more time, hopefully it passes this time :)

**Review by thiagocrepaldi — COMMENTED** [review]
One last failure

```bash
TestOnnxModelOutputConsistency_opset19CPU.test_output_match_atan_cpu_float16 (inputs='(tensor([[-2.9883, -4.0859,  5.5273,  1.6875, -1.6611, -2.2930,  6.2148,  3.1562,\n          0.0791,  6.3555, -6.5664,  6.6250,  7.0664, -3.7344,  0.6152,  4.1484,\n         -3.7266,  3.4805, -6.2422, -2.8047],\n        [ 4.2734, -4.1562, -8.2344, -7.4688,  5.2734, -1.5381,  5.9492, -4.2812,\n         -1.5732, -8.3672, -2.1719,  8.0469, -7.9883, -0.4219, -2.3633,  6.2305,\n          8.9844,  7.4453,  3.7891, -5.0625],\n        [-1.5293, -8.1328,  8.6484,  1.5557, -2.3633, -1.9951, -3.2070,  1.2920,\n         -1.0020, -5.2812,  2.5312,  8.4453,  2.3281, -2.8750, -3.3828, -4.2188,\n         -4.2266, -2.7246, -6.8555, -7.6719],\n        [-3.0664, -4.5156,  4.0352, -1.2480,  1.6084, -8.7500, -1.9512, -7.8750,\n         -7.8750,  8.1875,  0.4658,  7.9180,  4.8359,  4.2891, -2.3555, -6.6016,\n         -6.0547, -5.5898,  5.4219,  0.3691],\n        [ 3.8848, -4.2266, -1.1953, -6.7422,  2.8203,  8.7812,  8.2344, -8.5000,\n         -6.4414,  7.6484,  1.0459, -2.1523, -5.2188,  0.7207, -8.2031,  7.8750,\n         -5.5000, -3.6289,  4.9648, -1.0195],\n        [-3.4375,  8.2422, -8.2656, -7.2148, -6.1523, -2.1973, -0.5273,  0.4131,\n         -3.3828,  8.8906,  1.1514, -0.4570, -1.3711,  1.5820,  8.7891,  7.4375,\n          1.2568,  5.8008,  1.9951, -0.6152],\n        [-2.0469, -2.6797, -1.0107, -5.7227,  7.4961, -4.0781,  2.6797,  5.1055,\n         -7.5156,  5.9336,  3.9375,  0.8350,  1.7314,  5.3281, -6.7500,  7.7422,\n          6.0625,  2.5664,  8.1719, -1.8281],\n        [ 7.5664, -5.5977,  5.3594,  4.8672, -0.4482, -5.7656,  8.7812, -0.9404,\n         -7.2344, -5.6328,  4.5000,  7.9609, -5.7500, -4.1211, -0.6152,  3.3672,\n         -8.7969,  6.1445, -7.8594, -5.6875],\n        [-8.1406,  5.0469, -0.8174, -2.5488, -7.9180,  8.4609, -3.7793,  8.2812,\n         -2.5039,  1.3008, -7.1641, -4.0625,  3.6133, -2.1270,  0.3604,  0.5098,\n         -2.8047,  0.3164, -2.1621, -8.2422],\n        [ 0.3779, -1.3359, -8.9062, -2.2324, -6.3008,  0.7031, -8.7188,  1.5996,\n          1.1250, -1.9160, -6.5820,  1.5293,  4.8672,  3.2617,  1.6172, -5.0977,\n         -4.7109,  5.1758, -6.4844,  2.4609],\n        [-3.4023, -2.2324,  4.5273, -8.1719,  1.9863,  7.6914,  2.3633,  7.7344,\n         -6.3984,  2.8828,  0.1670, -3.6641,  7.3828,  3.6035, -5.0469,  2.1797,\n         -7.4180, -6.1602, -0.1758,  7.2852],\n        [ 6.8750,  7.5938,  2.2852,  8.4609, -2.4180, -7.7773, -0.0352, -3.7441,\n          4.0508, -1.1865,  5.4570, -1.5908, -2.8828,  4.6406,  1.3359, -0.8701,\n         -7.6914,  0.9668, -8.7812,  4.0781],\n        [-0.5977,  7.8125, -1.2129,  1.0283,  3.9102,  5.2305, -3.0859,  7.6641,\n         -0.8438,  2.0039,  4.4570, -2.2930, -7.9375,  2.8652, -4.8789, -6.4766,\n         -4.3867,  7.3750,  6.1172,  1.3359],\n        [-3.8047, -8.6406, -6.5938,  4.8516,  3.8496, -7.1797, -2.1523, -7.5391,\n         -8.0234,  3.0664,  8.9375,  1.1338, -2.5391,  7.7891, -1.8633, -7.1016,\n          3.3477, -8.7891, -8.4141, -0.4043],\n        [-0.1494,  6.7773, -6.9453,  6.2930,  8.7734, -6.4414,  6.3828, -0.0176,\n         -2.8398,  6.2305, -2.6094, -4.1211, -8.7188,  0.0967,  5.4141, -7.1367,\n          1.2129,  6.1016,  8.8125,  3.6992],\n        [ 5.5977,  5.8711, -3.5234,  4.2891, -2.2324,  4.7383,  8.6016,  1.9072,\n         -3.1719, -1.7842, -6.3008,  1.5293,  1.3008, -0.6592,  4.7539,  5.0898,\n         -4.1914, -1.9072,  3.4531, -8.3594],\n        [-6.3906, -0.9229,  7.9023,  7.1562, -2.5840,  8.3125, -1.3008,  6.7773,\n          7.0234, -2.1719,  3.4629,  4.7383, -5.2305, -7.2500,  3.7188,  0.5361,\n         -6.0039, -1.3975,  0.3867, -1.9512],\n        [ 8.1719,  0.7734, -8.7109, -3.3320, -6.9609, -2.2578, -8.6875,  6.7500,\n          4.0352, -6.4531, -7.2695, -1.1426,  8.3438,  2.2324,  0.7031,  3.9199,\n         -7.8047, -7.3398, -7.8672, -7.1641],\n        [ 3.2344, -6.4414,  3.3047,  3.0508, -7.5234,  7.7695, -1.9336, -7.4531,\n          4.7188, -6.1602,  7.6562, -4.5352, -1.9424,  1.0020,  7.0156,  0.2549,\n          6.6172,  2.5391, -7.7188,  2.3984],\n        [-7.0586, -6.0117, -4.5273, -4.4844, -8.3906,  3.1367, -1.0195, -4.1133,\n         -1.2217, -2.2773,  2.8301,  5.4766, -8.2031,  2.9961, -5.2812,  4.8438,\n          8.5312,  5.8203,  3.6387,  5.7734]], dtype=torch.float16),)', kwargs='{}', opset=19, sample_num=0) _
Traceback (most recent call last):
  File "onnx/test_op_consistency.py", line 307, in test_output_match
    self.run_test(model, inputs, rtol=rtol, atol=atol)
  File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 202, in run_test
    _run_test(model, tracing_remained_onnx_input_idx)
  File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 162, in _run_test
    return run_model_test(
  File "/var/lib/jenkins/workspace/test/onnx/onnx_test_common.py", line 82, in run_model_test
    return verification.verify(*args, options=options, **kwargs)
  File "<@beartype(torch.onnx.verification.verify) at 0x7f7d2202ddc0>", line 351, in verify
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/verification.py", line 890, in verify
    _compare_onnx_pytorch_model(
  File "<@beartype(torch.onnx.verification._compare_onnx_pytorch_model) at 0x7f7d2202bdc0>", line 159, in _compare_onnx_pytorch_model
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/verification.py", line 430, in _compare_onnx_pytorch_model
    onnx_session = _onnx_backend_session(onnx_model_f, options.backend)
  File "<@beartype(torch.onnx.verification._onnx_backend_session) at 0x7f7d2202b5e0>", line 52, in _onnx_backend_session
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/verification.py", line 227, in _onnx_backend_session
    onnx_session = _ort_session(model, (backend.value,))
  File "<@beartype(torch.onnx.verification._ort_session) at 0x7f7d2202b3a0>", line 63, in _ort_session
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/onnx/verification.py", line 197, in _ort_session
    ort_session = onnxruntime.InferenceSession(
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 383, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 435, in _create_inference_session
    sess.initialize_session(providers, provider_options, disabled_optimizers)
onnxruntime.capi.onnxruntime_pybind11_state.NotImplemented: [ONNXRuntimeError] : 9 : NOT_IMPLEMENTED : Failed to find kernel for Cast(19) (node InsertedCast_onnx::Atan_0). Op with name (InsertedCast_onnx::Atan_0) domain () and type (Cast) kernel is not supported in CPUExecutionProvider. Encountered following errors: ( Version mismatch. node_version: 19 kernel start version: 6 kernel_end_version: 12
 Version mismatch. node_version: 19 kernel start version: 13 kernel_end_version: 2147483647
```

**Comment by tcherckez-nvidia** [comment]
will upload a fix to the lint issue shortly

**Comment by tcherckez-nvidia** [comment]
@thiagocrepaldi  got my easyCLA approved, can you rerun the tests?

**Review by justinchuby — APPROVED** [review]
Thanks for the great work!

**Review by thiagocrepaldi — APPROVED** [review]

**Comment by thiagocrepaldi** [comment]
@malfet couold you look into this one? It adds 2 new dtypes to torchscript, needed by the onnx exporter

**Review by malfet — APPROVED** [review]

**Comment by kit1980** [comment]
This is failing internally in Meta with 

    error: no member named 'TensorProto_DataType_FLOAT8E5M2' in namespace 'onnx'
    error: no member named 'TensorProto_DataType_FLOAT8E4M3FN' in namespace 'onnx'
   
I believe this is because the version of onnx lib internally at meta doesn't have those.
You'll probably need to condition the code and only use that if available.

**Comment by thiagocrepaldi** [comment]
> This is failing internally in Meta with
> 
> ```
> error: no member named 'TensorProto_DataType_FLOAT8E5M2' in namespace 'onnx'
> error: no member named 'TensorProto_DataType_FLOAT8E4M3FN' in namespace 'onnx'
> ```
> 
> I believe this is because the version of onnx lib internally at meta doesn't have those. You'll probably need to condition the code and only use that if available.

this requires an onnx published after May 5, 2023 to work

**Comment by kit1980** [comment]
> This is failing internally in Meta with

And the diff was D48252254

**Comment by tcherckez-nvidia** [comment]
What should be fixed, the python tests? is there a compilation error?

**Comment by kit1980** [comment]
> What should be fixed, the python tests? is there a compilation error?

@tcherckez-nvidia Yes, compilation error.  As  I posted in https://github.com/pytorch/pytorch/pull/106379#issuecomment-1675189340

**Comment by tcherckez-nvidia** [comment]
@thiagocrepaldi @kit1980 I can repro the failing scenario when `third_party/onnx` is checked out to be an old version, prior to the introduction of fp8 to ONNX. 

However, I can't find a way to conditionally compile the code, as ONNX isn't providing any pre-processor flags that hint on the opset/ir version(I've spent effort searching for it and couldn't find, if you know how please share)

As I see it there are few options here:
1. Find a suitable pre-processor flag that ONNX supplies
2. Use a CPP compilation flag, which cmake will prepare after somehow realizing which ONNX lib is used for CPP compilation(not ONNX Python pkg) - I prefer to not go that way
3. Meta switch to use the open source ONNX version
4. Give up on fp8 ONNX export. I believe it has some benefit for the community but I'm not the one taking decisions
5. If you have other feasible suggestion, I'm ready to hear :)

btw, in bf16 ONNX export PR, there was no need for that kind of special conditioning - https://github.com/pytorch/pytorch/pull/66788

**Comment by tcherckez-nvidia** [comment]
6. another suggestion is to replace the enums with hard-coded values, 17 and 19

**Comment by tcherckez-nvidia** [comment]
@thiagocrepaldi let me know if you're onboard with suggestion 6

**Comment by justinchuby** [comment]
@thiagocrepaldi do you have thoughts? (6) sounds good to me. It would be nice if we can have this in pytorch.

**Comment by BowenBao** [comment]
Option 6 sounds good to me too. Let's unblock and land this PR.

@tcherckez-nvidia could you reopen this PR and make the necessary updates? Thanks!

**Comment by abock** [comment]
I have rebased the commits in this PR, resolved conflicts, defined the FP8 constants to avoid the build break, and have opened a new PR:

- [ ] #107962

FYI, The origin fork for this PR has been deleted, but GitHub preserved the commits.

**Comment by tcherckez-nvidia** [comment]
Thanks @abock

**Pull request closed without merging**
